### PR TITLE
inventory: add current pending requests

### LINKED ISSUE #2708: Search items by current number of pending requests
**Improvement suggestion**

As a librarian, I want to be able to search the most requested items for my library, so that I can manage my collections and periodically buy more copies of items that have many pending requests.

**Acceptance criteria**

- In the inventory list view: 
    - I can sort items by `current_pending_requests` (descending), so that I see most currently requested items at the top.
    - I can use an expert search to target for example: items with more than 5 current pending requests. 
    - The `current_pending_requests` is displayed in the listed items when it is not zero. 
    - The `current_pending_requests` is exported in the CSV of the inventory list
    - Nice-to-have: A range facet allows me to filter by number of pending requests.

**Use case explanation**

* In order to satisfy my patrons and avoid long waiting times for popular books, my library policy states that I need to buy one more copy of an item if it has more than X pending requests.
* As an acquisitions librarian, I need to stay on top of these most requested items. 
* Every week, I am able to view a list of the items that *currently* (at the time of my search) possess the highest number of *pending requests* so that I can buy more copies. I can then dispatch the pending requests to shorten the waiting time of my patrons.

### PR DESCRIPTION
**Pull request by Garfield-fr**
⚠️  Data Migration: Update mappings and reindex items.

* Closes #2708.

Co-Authored-by: Bertrand Zuchuat <bertrand.zuchuat@rero.ch>
Co-Authored-by: Peter Weber <peter.weber@rero.ch>

### CODE REVIEW

**Review by zannkukai — COMMENTED** [review]
I really not understand.
1) Facet is hidden and it's the main purpose of this issue
2) If a `loan.status` changed, how the `Item` index will be update ?
3) When the item will be reindex, the index chain isn't broken ; so each time a Loan resource will be indexed (each time loan.status changed!) you will index `Loan` itself (normal), related `Item` (Grrrrr but OK), but also `Holding` and `Document` (performance --) 
4) It's not necessary to reindex all items (seems huge job isn't it ?) but only items with current pending request (maybe I miss something) --> alembic script ?

**Comment by Garfield-fr** [comment]
> 1. Facet is hidden and it's the main purpose of this issue
In the US: Nice-to-have: A range facet allows me to filter by number of pending requests.
The range facet doesn't exists in NG-Core.

**Review by zannkukai — COMMENTED** [review]
Additionnaly, we already code a `get_request()` method on `Item` ressource. A 'request' isn't `PENDING` state. It also include `IN_TRANSIT_FOR_PICKUP` and `ITEM_AT_DESK`

**Comment by PascalRepond** [comment]
>     1. Facet is hidden and it's the main purpose of this issue

It is not the main purpose. Please read the issue more thouroughly.

**Comment by PascalRepond** [comment]
> Additionnaly, we already code a `get_request()` method on `Item` ressource. A 'request' isn't `PENDING` state. It also include `IN_TRANSIT_FOR_PICKUP` and `ITEM_AT_DESK`

We could use this existing function but it needs to be adapted. We don't have patron pid, only item.

https://github.com/rero/rero-ils/blob/8dc2b3fa5d1aca0391da743d02f608470ec6af4a/rero_ils/modules/loans/api.py#L1059-L1073

**Comment by zannkukai** [comment]
> We could use this existing function but it needs to be adapted. We don't have patron pid, only item.

https://github.com/rero/rero-ils/blob/staging/rero_ils/modules/items/api/circulation.py#L1209-L1218

File: rero_ils/config.py [hunk 1]
```diff
@@ -1766,6 +1766,9 @@ def _(x):
 # Default number of results in facet
 RERO_ILS_DEFAULT_AGGREGATION_SIZE = 30
 
+# current_pending_requests
+RERO_ILS_DEFAULT_CURRENT_PENDING_REQUESTS_INTERVAL = 1
```
Comment (jma) [inline rero_ils/config.py:1770]: Seems not used.

File: rero_ils/config.py [hunk 2]
```diff
@@ -1982,7 +1985,17 @@ def _(x):
                 terms=dict(
                     field='vendor.pid',
                     size=RERO_ILS_DEFAULT_AGGREGATION_SIZE)
-            )
+            ),
+            # TODO: Missing histogram facet in ng-core (and filter)
```
Comment (jma) [inline rero_ils/config.py:1989]: Do we want to keep this? It is not so difficult to rewrite later.

File: rero_ils/modules/items/api/circulation.py [hunk 3]
```diff
@@ -72,7 +72,8 @@ def change_status_commit_and_reindex(self):
         """
         current_search.flush_and_refresh(
             current_circulation.loan_search_cls.Meta.index)
-        self.status_update(self, dbcommit=True, reindex=True, forceindex=True)
+        self.status_update(
```
Comment (jma) [inline rero_ils/modules/items/api/circulation.py:75]: Only indentation has been touched in this file, am I right?

File: rero_ils/modules/items/serializers/collector.py [hunk 4]
```diff
@@ -195,7 +195,8 @@ def get_item_data(hit):
                 'temporary_item_type', {}).get('pid'),
             'item_masked': 'No',
             'item_status': hit.get('status'),
-            'issue': hit.get('issue', {})
+            'issue': hit.get('issue', {}),
+            'current_pending_requests': hit.get('current_pending_requests')
```
Comment (jma) [inline rero_ils/modules/items/serializers/collector.py:199]: You can perhaps return a default value: i.e. 0?

File: rero_ils/modules/loans/api.py [hunk 5]
```diff
@@ -71,6 +71,15 @@ class Meta:
 
         default_filter = None
 
+    def get_pending_request_count(self, item_pid):
```
Comment (jma) [inline rero_ils/modules/loans/api.py:74]: Interesting... I would put naturally put this in the Loan class, we have plenty of other example in the code. But why not in the Search. To be discuss with the team.

**Review by jma — APPROVED** [review]

File: rero_ils/modules/loans/api.py (hunk 5 above)
Comment (PascalRepond) [reply rero_ils/modules/loans/api.py:74]: This function already exists: 

https://github.com/rero/rero-ils/blob/staging/rero_ils/modules/items/api/circulation.py#L1209-L1218

File: rero_ils/modules/items/listener.py [hunk 6]
```diff
@@ -57,3 +59,7 @@ def enrich_item_data(sender, json=None, record=None, index=None,
         # Document type
         document = Document.get_record_by_pid(json['document']['pid'])
         json['document']['document_type'] = document['type']
+
+        # Current pending requests
+        json['current_pending_requests'] = \
+            LoansSearch().get_pending_request_count(item.pid)
```
Comment (zannkukai) [inline rero_ils/modules/items/listener.py:65]: `json['current_pending_requests '] = Item.get_requests(output='count')`

and you could remove your `LoansSearch().get_pending_request_count()` method.

**Review by zannkukai — APPROVED** [review]

File: rero_ils/modules/items/serializers/json.py [hunk 7]
```diff
@@ -106,4 +106,11 @@ def _postprocess_search_aggregations(self, aggregations: dict) -> None:
             aggregations.get('vendor', {}).get('buckets', []),
             VendorsSearch, 'name'
         )
+        if aggregations.get('current_pending_requests'):
+            aggregations['current_pending_requests']['type'] = 'range'
+            aggregations['current_pending_requests']['config'] = {
+                'min': 1,
+                'max': aggregations['current_pending_requests']['value'],
```
Comment (PascalRepond) [inline rero_ils/modules/items/serializers/json.py:113]: Nice ! :+1:

**Comment by PascalRepond** [comment]
## PO tests

- [ ] REP: Number of current requests line should be hidden in the list when it is zero
- [ ] REP: As mentionned in the ui PR : "I'd prefer "Current pending requests" for label. If it is too long, "Current requests" is better than "pending count"."
- [ ] REP: Small problem in the UI, Once I use the facet filter, I cannot change the filter unless from the facet unless I delete the filter.
- [ ] REP: The filter button could be more explicit : 
image

**Pull request merged**